### ISSUE TITLE
Discussion: versionning of po files

### ISSUE THREAD

**Issue by lingxiaoyang** [savoirfairelinux/sous-chef#703]
As one of developers, I find po files cumbersome when working with pull requests. Every time we run `manage.py makemessages`, Django updates the strings, locations and timestamp of every po file. Locations in particular -- a newline added somewhere in the code will trigger the update of the whole po file. It was sometimes a trouble when I tried to rebase my pull request or merge an other pull request, because it is so common to have various conflicts concerning po files and it takes too long time to resolve them.

Therefore, I would like to open a discussion on the versionning of po files for Sous-Chef. The basic idea comes from the GNU article [Put PO Files under Version Control](https://www.gnu.org/software/gettext/manual/html_node/Translations-under-Version-Control.html) that we may ignore the update of timestamp and locations in everyday development, so that we can make a better review on pull requests. When we prepare a release of Sous-Chef, however, we have to update the whole file in order to help translators on Transifex. How do you think of the idea? Please let me know.

**Comment by erozqba** [comment]
@lingxiaoyang :+1:

**Comment by lingxiaoyang** [comment]
I am proposing the following changes regarding our development workflow:

1. Drop the English `django.po` files. We don't work on these English po files as it is the default language for Sous-Chef. 
2. Add `--no-wrap` when generating po files. That is to say, don't break a translatable string into multiple lines. This will help the reading of pull request differences.
3. Drop locations of strings (i.e., from which file, in which line). 
   - Pros: we will avoid most changes on po files caused by an addition/deletion of code, and will speed up the manual merge of translation files. This is the only other option provided by `xgettext` that Django `makemessages` uses to extract translatable strings. We don't have the choice of displaying only the filenames but not line numbers.
   - Cons: it will take time to locate the origin of strings. From my own experience as a translator, it could be very difficult to translate short text (a word or two) without knowing the context. We have already split the po files by modules (member, order, meal, etc.), but still need to confirm with existing project translators on Transifex to see if they have had to consult the origin of some text.

In sum, this only needs to change the command that we use to generate po files. At present, we are running:

    python3 manage.py makemessages -l en -l fr

What I'm proposing is changing it to:

    python3 manage.py makemessages -l fr --no-wrap --no-location

And as for the one-time job, we update the existing French po files, delete English po files, commit and keep other developers synchronized.

**Comment by lingxiaoyang** [comment]
@madmath @PascalPriori  Hello Sous-Chef translators! I'd love to know during your work on Transifex, if you have ever consulted the origin of an English text (from which file, and in which line), in order to translate it into French.

This is because I'm proposing removing this information from our code repository for better version control at the developers' side, but I also have to know potential influences at your side. Do you think this would cause any problem during the translation process, according to your experience? Please let me know. Thank you!

**Comment by PascalPriori** [comment]
@lingxiaoyang Let's talk about it monday. I'm not sure to understand well.

**Comment by lingxiaoyang** [comment]
According to @PascalPriori, the location information of translatable strings is not crucial to translators, although it can be helpful considering translators work without the context of application. It is possible to remove this metadata and there will not be a major difficulty for them.

But just now, I found it possible to suppress the line numbers and keep only the file names in .po files. This feature was recently added to GNU gettext utilities in March 2014 (see commit 90c76fda268779b80c2fa969b8efddbb4aa5da1e in its Git repo) but not yet supported by Django makemessages command. I have submitted a feature request to django-developers group. I didn't find this feature before as it was seldom discussed and was not documented in `xgettext --help` or `man xgettext`.

In the meantime, I believe it is possible to extend the makemessages command by ourselves. I will work on it.

Other issues regarding the usage of Django translation, which we may consider to add in our docs:

1. Use `{% blocktrans trimmed %}` in templates to improve the readability of .po files. (Django doc reference: the paragraphs above [this section](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/#string-literals-passed-to-tags-and-filters).)

2. Add comments for translators when necessary (when we think it may be confusing without its context). Possible formats:
   - In Python: use [pgettext()](https://docs.djangoproject.com/en/1.10/ref/utils/#django.utils.translation.pgettext) or [translator comments](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/#comments-for-translators)
   - In templates: use [translator comments](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/#comments-for-translators-in-templates)
